### ISSUE TITLE
Port to Apache Cordova (phonegap)

### ISSUE THREAD

**Issue by jbtbnl** [owncloud-archive/windows8#5]
It would be great to port this application (the logic at least) to Apache Cordova to be able to compile it for all of the following platforms:

Android
Bada
Blackberry
FirefoxOS
iOS
Mac OS X
QT
Tizen
Ubuntu
WebOS
Windows (desktop)
Windows Phone 7
Windows Phone 8

It would enable us to start working on a single codebase for all ownCloud sync clients.

**Comment by riegerchris** [comment]
You can use the logic part (i.e. /lib) as is for further development for any platform as it contains pure javascript and was originally designed to work cross-platform (no guarantees but at least that was the intention).
In fact there was also a very early demonstration but far from anything near production.

**Comment by jbtbnl** [comment]
@MorrisJobke @jancborchardt @georgehrke would this be a feasible project for GSoC?

Taking the Win8 and FxOS apps as a starting point and build a cross-platform mobile app that has at least file-sync capabilities.

**Comment by MorrisJobke** [comment]
cc @cosenal

**Comment by jancborchardt** [comment]
Yeah, @cosenal & @Apophenia what do you think about expanding the photo upload app https://github.com/Apophenia/owncloud-photo-upload, or building upon it?

**Comment by cosenal** [comment]
I would say that first we should work on https://github.com/Apophenia/owncloud-photo-upload
It's not usable yet.

**Comment by jbtbnl** [comment]
@jancborchardt and @cosenal thanks for your replies, I believe it can be quite beneficial to transition to a cross-platform app because it will gather more attention from developers that way.

For example I'd happily do some work on a mobile app, but only if I can run it on my phone (being a BlackBerry). Furthermore it would prevent the community from basically reinventing the same app for every platform.

**Comment by cosenal** [comment]
@jbtbnl Of course I am in favor of this, my visionary past self (zimba12) had proposed this back in 2012! See https://mail.kde.org/pipermail/owncloud/2012-December/006979.html

**Comment by LukasReschke** [comment]
I wonder whether the visionary past you wants to help to implement this now ^_^

**Comment by cosenal** [comment]
@LukasReschke implementing??? pfft! that's for gsoc students, I am just the visionary here!

**Comment by jbtbnl** [comment]
@cosenal thanks for linking that discussion on the mailing list, I didn't know about it's existence :)

So do we agree that this might be a feasible project for GSoC? If so, I'll post it in the wiki.

**Comment by cosenal** [comment]
Well, the usual two concerns:
1. the project is clearly too big;
2. from previous gsoc/opw experience, it's better to start with something half-cooked by us on which the student can work.  
cc @jancborchardt

**Comment by Apophenia** [comment]
I think it sort of depends on how we want to prioritize this.

Personally, I would like to focus on finishing the Firefox OS app first, keeping in mind portability, so that we have one solid, working app that we know functions well on the FxOS platform._Then_ I would be more excited about porting it to Cordova in the form of a more general sync app. Of course, people are welcome to use my code in the meantime :)

However, I understand the impetus to pause the FxOS app for now and focus on using the existing code from FxOS and or/Windows to start a Cordova project, since in theory the Cordova code would be portable to FxOS anyway.

I think the way we pursue this might end up depending on how mature the cross-platform APIs in Cordova are. I have no doubt that Cordova is probably excellent for basic clicky-clicky things and CRUD apps, but I would be surprised if things like its camera, file system manipulation capabilities, and low-level networking capabilities were solid across all 13 platforms. Does anyone have experience with these parts of Cordova in particular?

**Comment by cosenal** [comment]
+1 to "one solid, working app that we know functions well on the FxOS platform". I am willing to help with that.

**Comment by jbtbnl** [comment]
@Apophenia just to be sure; I did not intend to say or imply that development on the FxOS app should be paused. The [platform support documentation](http://cordova.apache.org/docs/en/4.0.0/guide_support_index.md.html#Platform%20Support) of Cordova is pointing out that support for FxOS is currently not so good. It might be due to FxOS being relatively new or having a small userbase.

Keeping portability in mind during the development of the FxOS app would be a great start!
